common/mam/merkle.c

```diff
@@ -0,0 +1,195 @@
+#include "mam/merkle.h"
+#include "common/sign/v2/iss_curl.h"
+
+static trit_t const merkle_null_hash[HASH_LENGTH] = {0};
+
+// thanks to
+// https://stackoverflow.com/questions/3272424/compute-fast-log-base-2-ceiling#3272516
+static int ceil_log2(unsigned long long x) {
+  static const unsigned long long t[6] = {
+      0xFFFFFFFF00000000ull, 0x00000000FFFF0000ull, 0x000000000000FF00ull,
+      0x00000000000000F0ull, 0x000000000000000Cull, 0x0000000000000002ull};
+
+  int y = (((x & (x - 1)) == 0) ? 0 : 1);
+  int j = 32;
+  int depth_i;
+
+  for (depth_i = 0; depth_i < 6; depth_i++) {
+    int k = (((x & t[depth_i]) == 0) ? 0 : j);
+    y += k;
+    x >>= k;
+    j >>= 1;
+  }
+
+  return y;
+}
+
+static size_t binary_tree_size(size_t depth, size_t acc) {
+  if (depth == 0) {
+    return acc + 1;
+  }
+  return binary_tree_size(depth - 1, acc + (1 << depth));
+}
+
+static size_t merkle_node_base_index(size_t acc, size_t depth, size_t width,
+                                     size_t base_size) {
+  if (depth == 0) return acc;
+  acc += width + 1;
+  if (base_size % 2 == 0)
+    width = width / 2;
+  else
+    width = (width % 2 == 0 ? width / 2 : (width + 1) / 2);
+  return merkle_node_base_index(acc, depth - 1, width, (base_size + 1) / 2);
+}
+
+size_t merkle_node_index(size_t depth, size_t width, size_t base_size,
+                         size_t tree_depth) {
+  size_t acc = 0;
+  size_t idx = width;
+
+  while (depth < tree_depth) {
+    if (width != 0) idx = idx * 2 - 1;
+    acc--;
+    depth++;
+  }
+  return acc + merkle_node_base_index(0, tree_depth, idx, base_size);
+}
+
+size_t merkle_size(size_t leaf_count) {
+  size_t acc = 1;
+
+  while (leaf_count >= 2) {
+    acc += leaf_count;
+    leaf_count = (leaf_count + 1) >> 1;
+  }
+  return acc;
+}
+
+size_t merkle_depth(size_t node_count) {
+  size_t depth = 1;
+
+  while (binary_tree_size(depth, 0) < node_count) {
+    depth++;
+  }
+  return depth + 1;
+}
+
+int merkle_create(trit_t *const tree, trit_t *const seed, int64_t offset,
+                  const size_t base_size, size_t key_size, Curl *const c) {
+  size_t depth_i, e;
+  ssize_t j;
+  trit_t key[key_size];
+
+  const size_t td = ceil_log2(base_size);
+
+  // create addresses
+  for (depth_i = 0; depth_i < base_size; depth_i++) {
```

Comment:
depth_i == leaf_index ?
if so, it's clearer to rename to something indicating we are traversing horizontally rather than vertically..

Reply:
Will do, thanks.